File: 1219. Path with Maximum Gold.java
class Solution {
    public int getMaximumGold(int[][] grid) {
        int max = 0;
        for(int x = 0; x < grid.length; x++) {
            for(int y = 0; y < grid[0].length; y++) {
                max = Math.max(dfs(grid, x , y, 0), max);
            }
        }
        return max;
    }

    public int dfs(int[][] grid, int x, int y, int sum) {
        if(grid.length-1 < x || grid[0].length-1 < y || x < 0 || y < 0) return sum;
        if(grid[x][y] == 0) return sum;
        int[][] arrow = new int[][] {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        int max = sum;
        for(int i = 0 ; i< arrow.length; i++) {
            int nextX = arrow[i][0] + x;
            int nextY = arrow[i][1] + y;
            int temp = grid[x][y];
            grid[x][y] = 0;
            max = Math.max(dfs(grid, nextX, nextY, sum + temp), max);
            grid[x][y] = temp;
        }
        return max;
    }
}

File: 1219. Path with Maximum Gold.kt
class Solution {
    var arrow = arrayOf(1 to 0, 0 to 1, -1 to 0, 0 to -1)

    // 최대 골드를 얻는 함수
    fun getMaximumGold(grid: Array<IntArray>): Int {
        var result = 0
        for (x in grid.indices) {
            for (y in grid[0].indices) {
                // 골드가 있는 셀에서만 DFS를 시작
                if (grid[x][y] > 0) {
                    result = maxOf(result, dfs(x, y, 0, grid))
                }
            }
        }
        return result
    }
    private fun dfs(x: Int, y: Int, sum: Int, grid: Array<IntArray>): Int {
        // 경계 조건 및 방문 조건 검사
        if (x < 0 || y < 0 || x >= grid.size || y >= grid[0].size || grid[x][y] <= 0) {
            return sum
        }

        // 현재 셀의 골드를 임시로 저장하고 방문 표시
        val temp = grid[x][y]
        grid[x][y] = 0

        var max = sum
        // 네 방향으로 DFS 호출
        for ((dx, dy) in arrow) {
            max = maxOf(max, dfs(x + dx, y + dy, sum + temp, grid))
        }

        // 방문 표시 해제 (백트래킹)
        grid[x][y] = temp
        return max
    }
}
